aten/src/ATen/native/UpSample.h

```diff
@@ -467,30 +469,32 @@ static inline void compute_source_index_and_lambda(
   }
 }
 
-// It will not be used by data types other than BFloat16.
-template <typename scalar_in, typename scalar_out>
+// It will not be used by data types other than BFloat16 and Half.
+template <typename scalar_in, typename scalar_out,
+          typename std::enable_if_t<!vec::is_reduced_floating_point_v<scalar_out> || !std::is_same<scalar_in, float>::value, int> = 0>
 void inline apply_grad_input(scalar_in* buffer_ptr, scalar_out* gin, int64_t size) {
-  TORCH_CHECK((std::is_same<scalar_out, BFloat16>::value),
-              "Upsample backward only support BFloat16 in the lower percision data types on CPU.")
+  TORCH_CHECK((vec::is_reduced_floating_point_v<scalar_out>),
+              "Upsample backward only support BFloat16 and Half in the lower percision data types on CPU.")
   TORCH_CHECK((std::is_same<scalar_in, float>::value),
-              "Upsample backward should use float as acc buffer for BFloat16 grad input on CPU.")
+              "Upsample backward should use float as acc buffer for BFloat16 and Half grad input on CPU.")
   return;
 }
 
-template <>
-void inline apply_grad_input(float* buffer_ptr, BFloat16* gin, int64_t size) {
-  using bVec = vec::Vectorized<BFloat16>;
+template <typename scalar_in, typename scalar_out,
+          typename std::enable_if_t<vec::is_reduced_floating_point_v<scalar_out> && std::is_same<scalar_in, float>::value, int> = 0>
+void inline apply_grad_input(scalar_in* buffer_ptr, scalar_out* gin, int64_t size) {
+  using bVec = vec::Vectorized<scalar_out>;
   using fVec = vec::Vectorized<float>;
   int64_t d = 0;
   for (; d < size - (size % bVec::size()); d += bVec::size()) {
     bVec gin_bvec = bVec::loadu(gin + d);
     fVec gin_fvec0, gin_fvec1;
-    std::tie(gin_fvec0, gin_fvec1) = convert_bfloat16_float(gin_bvec);
+    std::tie(gin_fvec0, gin_fvec1) = vec::convert_to_float<scalar_out>(gin_bvec);
```

Comment:
why do we need the namespace of `vec` here ?

Reply:
I added the line "using namespace vec;" at the beginning, and remove the namespace of vec here.